### PR TITLE
ODC-7313: Run CI tests with non-admin user

### PR DESCRIPTION
**Pull request by sanketpathak**
Issue: https://issues.redhat.com/browse/ODC-7313

Description: Running pull CI tests for devconsole, pipelines, and knative packages with non-admin user

### CODE REVIEW

File: frontend/packages/dev-console/integration-tests/support/pages/app.ts [hunk 1]
```diff
@@ -262,12 +263,24 @@ export const projectNameSpace = {
     app.waitForLoad();
     cy.url().then((url) => {
       if (url.includes('add/all-namespaces')) {
-        cy.get('tr[data-test-rows="resource-row"]').should('have.length.at.least', 1);
+        cy.get(userPreferencePO.userMenu, {
+          timeout: 50000,
```
Comment (jerolimov) [inline frontend/packages/dev-console/integration-tests/support/pages/app.ts:267]: 50 seconds? :smirk:

File: frontend/packages/dev-console/integration-tests/testData/create_user.sh [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    htpasswd:
+      fileData:
+        name: ${HTPASSWD_SECRET}
+EOF
+
+sleep 10s
+oc create clusterrolebinding ${USERNAME}_role1 --clusterrole=developer --user=${USERNAME}
+oc create clusterrolebinding ${USERNAME}_role2 --clusterrole=self-provisioner --user=${USERNAME}
+oc create clusterrolebinding ${USERNAME}_role2 --clusterrole=basic-user --user=${USERNAME}
+sleep 15s
```
Comment (jerolimov) [inline frontend/packages/dev-console/integration-tests/testData/create_user.sh:40]: Can we wait here on an console restart as we had saw this together in the `test-prow-e2e.sh` script?

File: frontend/packages/knative-plugin/integration-tests/support/commands/hooks.ts [hunk 3]
```diff
@@ -7,6 +7,7 @@ import {
   navigateTo,
   perspective,
 } from '@console/dev-console/integration-tests/support/pages';
+// import { userLoginPage } from '../pages/dev-perspective/common';
```
Comment (jerolimov) [inline frontend/packages/knative-plugin/integration-tests/support/commands/hooks.ts:10]: Shouldn't we remove unused code?

File: frontend/packages/knative-plugin/integration-tests/support/commands/hooks.ts [hunk 4]
```diff
@@ -23,6 +24,7 @@ before(() => {
     }
     cy.log('uncaught:exception', err, runnable, promise);
   });
+  // userLoginPage.nonAdminUserlogin();
```
Comment (jerolimov) [inline frontend/packages/knative-plugin/integration-tests/support/commands/hooks.ts:27]: Shouldn't we remove unused code?

File: frontend/packages/knative-plugin/integration-tests/support/commands/hooks.ts (hunk 3 above)
Comment (sanketpathak) [reply frontend/packages/knative-plugin/integration-tests/support/commands/hooks.ts:10]: Just tried something, will remove in the final cut

File: frontend/packages/dev-console/integration-tests/support/pages/app.ts (hunk 1 above)
Comment (jerolimov) [reply frontend/packages/dev-console/integration-tests/support/pages/app.ts:267]: But 50 seconds? I would hope our user menu never takes so long to render...?

File: frontend/packages/dev-console/integration-tests/testData/create_user.sh [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+      fileData:
+        name: ${HTPASSWD_SECRET}
+EOF
+
+sleep 10s
+oc create clusterrolebinding ${USERNAME}_role1 --clusterrole=developer --user=${USERNAME}
+oc create clusterrolebinding ${USERNAME}_role2 --clusterrole=self-provisioner --user=${USERNAME}
+oc create clusterrolebinding ${USERNAME}_role2 --clusterrole=basic-user --user=${USERNAME}
+
+  output=''
```
Comment (jerolimov) [inline frontend/packages/dev-console/integration-tests/testData/create_user.sh:41]: Intention isn't right

File: frontend/packages/dev-console/integration-tests/testData/create_user.sh [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+  count=0
+  sleepDuration=15
+  maxRetries=40  #10 minutes
+  action="started"
+  if [ "$1" == "False" ]; then
```
Comment (jerolimov) [inline frontend/packages/dev-console/integration-tests/testData/create_user.sh:46]: Multiple times this code mentions `$1`. I think this is a left over because it was a function before, or?

File: frontend/packages/knative-plugin/integration-tests/support/step-definitions/common/common.ts [hunk 7]
```diff
@@ -22,9 +23,20 @@ import {
   verifyAndInstallKnativeOperator,
   createChannel,
   verifyAndInstallOperator,
+  app,
 } from '@console/dev-console/integration-tests/support/pages';
 import { eventingPO } from '@console/knative-plugin/integration-tests/support/pageObjects/global-po';
 
+Given('user has logged in as a basic user', () => {
+  cy.logout();
```
[the remark this answers is not in the document]
Comment (sanketpathak) [reply frontend/packages/knative-plugin/integration-tests/support/step-definitions/common/common.ts]: Made the change

File: frontend/packages/dev-console/integration-tests/testData/create_user.sh (hunk 6 above)
Comment (sanketpathak) [reply frontend/packages/dev-console/integration-tests/testData/create_user.sh]: Removed it

File: frontend/packages/dev-console/integration-tests/testData/create_user.sh (hunk 5 above)
Comment (sanketpathak) [reply frontend/packages/dev-console/integration-tests/testData/create_user.sh]: Removed it

File: frontend/packages/knative-plugin/integration-tests/support/commands/hooks.ts (hunk 4 above)
Comment (sanketpathak) [reply frontend/packages/knative-plugin/integration-tests/support/commands/hooks.ts]: Removed it

**Comment by sanketpathak** [comment]
/retitle [ODC-7313](https://issues.redhat.com/browse/ODC-7313): Run CI tests with non-admin user

File: frontend/packages/dev-console/integration-tests/testData/create_user.sh [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+
+sleep 10s
+oc create clusterrolebinding ${USERNAME}_role1 --clusterrole=developer --user=${USERNAME}
+oc create clusterrolebinding ${USERNAME}_role2 --clusterrole=self-provisioner --user=${USERNAME}
+oc create clusterrolebinding ${USERNAME}_role2 --clusterrole=basic-user --user=${USERNAME}
+
+until [ $count -gt $maxRetries ]; do
+  ((count=count+1))
+  sleep $sleepDuration
+done
```
Comment (jerolimov) [inline frontend/packages/dev-console/integration-tests/testData/create_user.sh:48]: Wait... This just waits.. There no extra condition to break this loop other than the count? This is different then https://github.com/openshift/console/blob/b73f9bd3f0e8626b32939707b096dac9b624187b/test-prow-e2e.sh#L26

So this will always 10 minutes (maxRetries 40 x sleepDuration 15 = 600 seconds).

File: frontend/packages/dev-console/integration-tests/testData/create_user.sh [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+    type: HTPasswd
+    htpasswd:
+      fileData:
+        name: ${HTPASSWD_SECRET}
+EOF
+
+sleep 10s
+oc create clusterrolebinding ${USERNAME}_role1 --clusterrole=developer --user=${USERNAME}
+oc create clusterrolebinding ${USERNAME}_role2 --clusterrole=self-provisioner --user=${USERNAME}
+oc create clusterrolebinding ${USERNAME}_role2 --clusterrole=basic-user --user=${USERNAME}
```
[the remark this answers is not in the document]
Comment (sanketpathak) [reply frontend/packages/dev-console/integration-tests/testData/create_user.sh:43]: It used to work before

**Review by jerolimov — APPROVED** [review]
This lgtm if the e2e test passes on the CI. I will also try it on my machine later and would add lgtm then.

Any objection @spadgett?

File: frontend/packages/dev-console/integration-tests/support/commands/hooks.ts [hunk 10]
```diff
@@ -1,9 +1,13 @@
+import { guidedTour } from '@console/cypress-integration-tests/views/guided-tour';
+
 before(() => {
-  const bridgePasswordIDP: string = Cypress.env('BRIDGE_HTPASSWD_IDP');
-  const bridgePasswordUsername: string = Cypress.env('BRIDGE_HTPASSWD_USERNAME');
-  const bridgePasswordPassword: string = Cypress.env('BRIDGE_KUBEADMIN_PASSWORD');
+  cy.exec(`../../../../create-user.sh`, { failOnNonZeroExit: false });
```
Comment (jerolimov) [inline frontend/packages/dev-console/integration-tests/support/commands/hooks.ts:4]: Hey Sanket, this was the part we didn't test on Friday. But this "do not fail" (failOnNonZeroExit) makes me worried, so I run the tests manually and looked into the details.

TL;DR:

```suggestion
  cy.exec('../../../../contrib/create-user.sh');
```

How I come and verified this:

I inspected the exec command result in cypress and opened the browser log (F12) to see all the details when I clicked then on exec:

Screenshot from 2023-07-02 19-32-22

I dropped failOnNonZeroExit to let the test fail and get the full error in the sidebar:

Screenshot from 2023-07-02 19-33-49

Because I didn't saw directly the error I changed the command to `pwd`:

Screenshot from 2023-07-02 19-34-32

Then I understand that the `contrib` folder was missing. After fixing the path I got:

Screenshot from 2023-07-02 19-37-38

So the create-user.sh script was started correctly. But it couldn't load the `htpasswd-secret.yaml` because the cwd was the integration-test folder instead of the console (root) folder.

For testing I removed the debug (`x`) option from `set -ex..`

Screenshot from 2023-07-02 19-40-36

Then I tested the script from the terminal to reproduce the same error...

Screenshot from 2023-07-02 19-42-48

Screenshot from 2023-07-02 19-43-00

File: contrib/create-user.sh [hunk 11]
```diff
@@ -0,0 +1,45 @@
+#!/usr/bin/env bash
+
+set -exuo pipefail
+
+function waitForAuthOperatorProgressing {
```
Comment (jerolimov) [inline contrib/create-user.sh:5]: To allow that this script is called from different places, we need to get the root or base folder (cloned console folder).

You can test this by running the `create-user.sh` script from different folders. For example

from console folder:
```
contrib/create-user.sh
```

or from the integration-test folder:

```
cd frontend/packages/dev-console/integration-tests
../../../../contrib/create-user.sh
```

```suggestion
set -exuo pipefail

BASEDIR=$(dirname "$0")/..

function waitForAuthOperatorProgressing {
```

File: contrib/create-user.sh [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+}
+
+# Add htpasswd IDP
+# Skip wait for authentication operator to start Progressing when the Secret already exists.
+# And ignore any kind of errors when the Secret doesn't exist.
+htpasswd_secret_name=$(oc get --ignore-not-found secret -n openshift-config htpass-secret -o name 2> /dev/null)
+
+if [ "$htpasswd_secret_name" == "" ]; then
+  oc create -f ./frontend/integration-tests/data/htpasswd-secret.yaml
+  oc patch oauths cluster --patch "$(cat ./frontend/integration-tests/data/patch-htpasswd.yaml)" --type=merge
```
Comment (jerolimov) [inline contrib/create-user.sh:38]: Use the BASEDIR to load the two yamls files from different working dirs:

```suggestion
  oc create -f "$BASEDIR//frontend/integration-tests/data/htpasswd-secret.yaml"
  oc patch oauths cluster --patch "$(cat "$BASEDIR/frontend/integration-tests/data/patch-htpasswd.yaml")" --type=merge
```

**Review by jerolimov — CHANGES_REQUESTED** [review]
Hey, I tested this and found two issues:

1. After running the `create-user.sh` script I can not login as kubeadmin on my CRC... Need to test this on a cluster bot as well...
   * **update:** On a cluster bot instance I could login with kubeadmin and test after the script finishes.
3. I noticed that the `create-user.sh` script isn't executed when I run `yarn test-cypress-dev-console` and started the add-page tests. See below:

**Review by jerolimov — APPROVED** [review]
Code looks good to me. Thanks for making this possible :+1:

In the CI job https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_console/12814/pull-ci-openshift-console-master-e2e-gcp-console/1675927075862614016 was a pipeline-ci.mp4 where I could also confirm that the tests installs the Pipelines operator as kubeadmin and then switches to the test user:

image

/lgtm
/approve

**Comment by jerolimov** [comment]
chore: Only touches our e2e tests!

/label docs-approved
/label px-approved
/label qe-approved
